panther_analysis_tool/testing.py

```diff
@@ -0,0 +1,240 @@
+    output: Optional[str]
+    # error contains a TestError instance with the error message or
+    # None if no error was raised
+    error: Optional[TestError]
+
+    @classmethod
+    def new(
+        cls,
+        output: Optional[Union[bool, str, List[str]]],
+        raw_exception: Optional[Exception] = None,
+    ) -> Optional["FunctionTestResult"]:
+        """Create a new instance while applying
+        the necessary transformations to the parameters"""
+        if output is None and raw_exception is None:
+            return None
+
+        if output is not None and not isinstance(output, str):
+            output = json.dumps(output)
+
+        return cls(
+            output=output,
+            error=cls.to_test_error(raw_exception),
+        )
+
+    @staticmethod
+    def format_exception(exc: Optional[Exception], title: Optional[str] = None) -> Optional[str]:
+        """Convert an exception instance to a structured error message"""
+        if exc is None:
+            return None
+        if title is not None:
+            prefix = f"{title}: "
+        else:
+            prefix = ""
+
+        return f"{prefix}{type(exc).__name__}: {exc}"
+
+    @staticmethod
+    def to_test_error(exc: Optional[Exception], title: Optional[str] = None) -> Optional[TestError]:
+        """Convert an exception instance to a TestError,
+        also properly formatting the exception message"""
+        if exc is None:
+            return None
+        return TestError(message=FunctionTestResult.format_exception(exc, title=title))
+
+
+@dataclass  # pylint: disable=R0902
+class TestResultsPerFunction:
+    """Container for the results of each function"""
+
+    detectionFunction: Optional[FunctionTestResult]  # pylint: disable=C0103
+    titleFunction: Optional[FunctionTestResult] = None  # pylint: disable=C0103
+    dedupFunction: Optional[FunctionTestResult] = None  # pylint: disable=C0103
+    alertContextFunction: Optional[FunctionTestResult] = None  # pylint: disable=C0103
+    descriptionFunction: Optional[FunctionTestResult] = None  # pylint: disable=C0103
+    referenceFunction: Optional[FunctionTestResult] = None  # pylint: disable=C0103
+    severityFunction: Optional[FunctionTestResult] = None  # pylint: disable=C0103
+    runbookFunction: Optional[FunctionTestResult] = None  # pylint: disable=C0103
+    destinationsFunction: Optional[FunctionTestResult] = None  # pylint: disable=C0103
+
+
+@dataclass  # pylint: disable=R0902
+class TestResult:
+    """The structure of the results for a test case evaluation"""
+
+    id: Optional[str]  # pylint: disable=C0103
+    name: str
+    # The following two fields do not conform to Python's naming
+    # conventions, but the field names correspond
+    # to response attributes by API & FE.
+    # TODO: provide a field name translation step if necessary
+    detectionId: Optional[str]  # pylint: disable=C0103
+    genericError: Optional[str]  # pylint: disable=C0103
+    # TODO:
```

Comment:
do we still need this TODO? not sure what it means if we do

Reply:
haha - nope! I'll remove